In [2]:
#Downloading the file using urllib
import urllib.request
url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'
d_file = urllib.request.urlretrieve(url, 'covid_data.csv')

Downloading file
Using pandas to read file
finding info, describe data set, view columns,
Retrieving data row first,then colum
Retriveing data column first then row
How to aggregate columns (group data)
How to get a copy of the data
Retrieving a row or range of rows
Printing the beginner and ending rows of a dataset
Print out a series
Finding the first non-empty index in a series

In [3]:
#Using read_csv to read data
import pandas as pd

data_file = pd.read_csv("covid_data.csv")

In [4]:
data_file

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [9]:
data_file['new_cases'][3]

0.0

In [5]:
#Using .info to get a basic over view of the data
data_file.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [6]:
#Using .decribe() to get std, number of non-nul values, means...etc, max(highest value in that column), min(highest value in that column)
data_file.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

In [10]:
#.columns For list of headers/columns
data_file.columns



Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [13]:
#Retrieve number of rows and columns using .shape
data_file.shape

(248, 4)

In [5]:
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

In [10]:
y = len(covid_data_dict['date'])
final_list = []
p = 0
for i in range(y):
    f = []
    for key in covid_data_dict.keys():
        f.append(covid_data_dict[key][p])
    p = p + 1
    print(f)
    
    final_list.append(f)
    
print(final_list)


['2020-08-30', 1444, 1, 53541]
['2020-08-31', 1365, 4, 42583]
['2020-09-01', 996, 6, 54395]
['2020-09-02', 975, 8, None]
['2020-09-03', 1326, 6, None]
[['2020-08-30', 1444, 1, 53541], ['2020-08-31', 1365, 4, 42583], ['2020-09-01', 996, 6, 54395], ['2020-09-02', 975, 8, None], ['2020-09-03', 1326, 6, None]]


In [39]:
#Just me trying smth here
r = len(covid_data_dict['date'])

In [40]:
#Just me trying smth here

x = 0
master_list = []
for i in range(r):
    j = []
    for i in covid_data_dict.keys(): 
        j.append(covid_data_dict[i][x])
    master_list.append(j)
    print(j)
    x += 1
    
print(master_list)    

['2020-08-30', 1444, 1, 53541]
['2020-08-31', 1365, 4, 42583]
['2020-09-01', 996, 6, 54395]
['2020-09-02', 975, 8, None]
['2020-09-03', 1326, 6, None]
[['2020-08-30', 1444, 1, 53541], ['2020-08-31', 1365, 4, 42583], ['2020-09-01', 996, 6, 54395], ['2020-09-02', 975, 8, None], ['2020-09-03', 1326, 6, None]]


In [4]:
#Based on covid_data_dict Retrieveing information

covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

In [9]:
data_file

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [12]:
#.at method to retrieve the element at a specific row & column directly.
data_file.at[246, 'new_cases']

975.0

In [14]:
#Either you use the .at, or use this indexing method here column before row.
data_file['new_cases'][246]

975.0

In [15]:
#If a column does not conatin spaces or special characters you can call it using the . notation 
data_file.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

In [20]:
#To get a group of columns
case_death = data_file[['new_cases','new_tests']]

case_death

new_cases  new_tests
0          0.0        NaN
1          0.0        NaN
2          0.0        NaN
3          0.0        NaN
4          0.0        NaN
..         ...        ...
243     1444.0    53541.0
244     1365.0    42583.0
245      996.0    54395.0
246      975.0        NaN
247     1326.0        NaN

[248 rows x 2 columns]

In [21]:
#We can use .copy() to create a copy of the existing dataframe .The data within .copy() is completely separate from original, and changing values inside one of them will not affect the other.
data_file_copy = data_file.copy()
data_file_copy

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [23]:
#To get specific role .loc()
data_file.loc[244]

date          2020-08-31
new_cases         1365.0
new_deaths           4.0
new_tests        42583.0
Name: 244, dtype: object

In [24]:
#.head() , .tails() to print the first few andf last few rows of a data
data_file.head()

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN
3  2020-01-03        0.0         0.0        NaN
4  2020-01-04        0.0         0.0        NaN

In [26]:
data_file.tail()

date  new_cases  new_deaths  new_tests
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

In [27]:
#Finding the first non Nan value using .first_valid_index(), zero is a value Nan is no value
data_file.first_valid_index()

0

In [34]:
#Let's verify this information
data_file.loc[0:10]

date  new_cases  new_deaths  new_tests
0   2019-12-31        0.0         0.0        NaN
1   2020-01-01        0.0         0.0        NaN
2   2020-01-02        0.0         0.0        NaN
3   2020-01-03        0.0         0.0        NaN
4   2020-01-04        0.0         0.0        NaN
5   2020-01-05        0.0         0.0        NaN
6   2020-01-06        0.0         0.0        NaN
7   2020-01-07        0.0         0.0        NaN
8   2020-01-08        0.0         0.0        NaN
9   2020-01-09        0.0         0.0        NaN
10  2020-01-10        0.0         0.0        NaN

Summary 

covid_df['new_cases'] - Retrieving columns as a Series using the column name
new_cases[243] - Retrieving values from a Series using an index
covid_df.at[243, 'new_cases'] - Retrieving a single value from a data frame
covid_df.copy() - Creating a deep copy of a data frame
covid_df.loc[243] - Retrieving a row or range of rows of data from the data frame
head, tail, and sample - Retrieving multiple rows of data from the data frame
covid_df.new_tests.first_valid_index - Finding the first non-empty index in a series